### PR TITLE
Implement gdk-pixbuf loader

### PR DESCRIPTION
**Pull request by ohwgiles**
Providing a gdk-pixbuf loader module means applications such as `gpicview` and `pcmanfm` automatically gain HEIF support.

There are probably some improvements to be made, it would be especially nice to use the embedded thumbnails, but application code doesn't seem to reliably indicate to gdk-pixbuf that a thumbnail could be returned

### CODE REVIEW

**Comment by farindk** [comment]
Thank you, this is great.

@fancycode : I guess we will have to put this into a separate package in order to remove the dependency on gdk-pixbuf for the main library. Can you have a look how to do this?

Note: we might also need a post-install script that runs

    cache_file=`pkg-config gdk-pixbuf-2.0 --variable gdk_pixbuf_cache_file`
    gdk-pixbuf-query-loaders >$cache_file

to immediately update the cache of existing loaders.

**Comment by ohwgiles** [comment]
Thanks for the positive reception. I've added a commit to optionally compile the loader. I left it to default ON but maybe you want to change that.

I'm not sure I can help with the autoconf.

Regarding the post-install script to update the cache, I suggest this is better left to distributions, since as soon as you build deb/rpm/etc it has to be done differently, not to mention when cross-compiling.

**Comment by fancycode** [comment]
> Regarding the post-install script to update the cache, I suggest this is better left to distributions, since as soon as you build deb/rpm/etc it has to be done differently, not to mention when cross-compiling.

I agree, this is something that should be done by the packaging.

**Comment by fancycode** [comment]
> @fancycode : I guess we will have to put this into a separate package in order to remove the dependency on gdk-pixbuf for the main library. Can you have a look how to do this?

Similar as we do with the thumbnailer: the package `libgdk-pixbuf2.0-dev` will be added as build-dependency and the resulting library will be put in a separate binary package that then has a dependency on the `libheif` package, `libgdk-pixbuf2.0-0` and whatever else will be pulled in during linking. This (new) binary package will also run the post-install script to update the cache.

**Comment by farindk** [comment]
Yes, sounds good.

File: gdk-pixbuf/pixbufloader-heif.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
+/*
+ * gdk-pixbuf loader module for libheif
+ * Copyright (c) 2019 Oliver Giles <ohw.giles@gmail.com>
+ *
+ * This file is part of libheif.
+ *
+ * libheif is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
```
[the remark this answers is not in the document]
Comment (ohwgiles) [reply gdk-pixbuf/pixbufloader-heif.c]: My mistake - I intended LGPLv3. Fix pushed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,3 +63,8 @@ endif()
 
 add_subdirectory (examples)
 add_subdirectory (libheif)
+
+option(WITH_GDK_PIXBUF_LOADER "Compile a gdk-pixbuf loader plugin" ON)
```
[the remark this answers is not in the document]
Comment (ohwgiles) [reply CMakeLists.txt]: Well yes, but IMO it's nice to always be explicit about desired features and then use `find_package(.. REQUIRED)`. Anyway this is also fine.

File: gdk-pixbuf/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(GDKPIXBUF2 REQUIRED gdk-pixbuf-2.0)
+execute_process(COMMAND ${PKG_CONFIG_EXECUTABLE} gdk-pixbuf-2.0 --variable gdk_pixbuf_moduledir OUTPUT_VARIABLE GDKPIXBUF2_MODULE_DIR OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+add_library(pixbufloader-heif MODULE pixbufloader-heif.c)
+
+target_compile_definitions(pixbufloader-heif PRIVATE "-DGDK_PIXBUF_ENABLE_BACKEND")
```
[the remark this answers is not in the document]
Comment (ohwgiles) [reply gdk-pixbuf/CMakeLists.txt]: :+1:

**Comment by fancycode** [comment]
Thanks for your contribution!

**Pull request merged**